google/cloud/storage/internal/curl_download_request.cc

```diff
@@ -0,0 +1,163 @@
+      });
+  handle_.SetHeaderCallback([this](char* contents, std::size_t size,
+                                   std::size_t nitems) {
+    return CurlAppendHeaderData(
+        received_headers_, static_cast<char const*>(contents), size * nitems);
+  });
+  handle_.EnableLogging(logging_enabled_);
+}
+
+std::size_t CurlDownloadRequest::WriteCallback(void* ptr, std::size_t size,
+                                               std::size_t nmemb) {
+  handle_.FlushDebug(__func__);
+  GCP_LOG(DEBUG) << __func__ << "() size=" << size << ", nmemb=" << nmemb
+                 << ", buffer.size=" << buffer_.size();
+
+  buffer_.append(static_cast<char const*>(ptr), size * nmemb);
+  if (buffer_.size() > 128 * 1024) {
+    return CURL_READFUNC_PAUSE;
+  }
+  return size * nmemb;
+}
+
+int CurlDownloadRequest::PerformWork() {
+  // Block while there is work to do, apparently newer versions of libcurl do
+  // not need this loop and curl_multi_perform() blocks until there is no more
+  // work, but is it pretty harmless to keep here.
+  int running_handles = 0;
+  CURLMcode result;
+  do {
+    result = curl_multi_perform(multi_.get(), &running_handles);
+    GCP_LOG(DEBUG) << __func__ << "(): running_handles=" << running_handles
+                   << ", result=" << result;
+  } while (result == CURLM_CALL_MULTI_PERFORM);
+
+  // Raise an exception if the result is unexpected, otherwise return.
+  HandleCurlMultiErrorCode(__func__, result);
+  if (running_handles == 0) {
+    // The only way we get here is if the handle "completed", and therefore the
+    // transfer either failed or was successful. Pull all the messages out of
+    // the info queue until we get the message about our handle.
+    int remaining;
+    do {
+      auto msg = curl_multi_info_read(multi_.get(), &remaining);
+      if (msg->easy_handle != handle_.handle_.get()) {
+        // Raise an exception if the handle is not the right one. This should
+        // never happen, but better to give some meaningful error in this case.
+        std::ostringstream os;
+        os << __func__ << " unknown handle returned by curl_multi_info_read()"
+           << ", msg.msg=[" << msg->msg << "]"
+           << ", result=[" << msg->data.result
+           << "]=" << curl_easy_strerror(msg->data.result);
+        google::cloud::internal::RaiseRuntimeError(os.str());
+      }
+      GCP_LOG(DEBUG) << __func__ << "(): msg.msg=[" << msg->msg << "], "
+                     << " result=[" << msg->data.result
+                     << "]=" << curl_easy_strerror(msg->data.result);
+      // The transfer is done, set the state flags appropriately.
+      curl_closed_ = true;
+    } while (remaining > 0);
+  }
+  return running_handles;
+}
+
+void CurlDownloadRequest::WaitForHandles() {
+  CURLMcode result = curl_multi_wait(multi_.get(), nullptr, 0, 0, nullptr);
+  HandleCurlMultiErrorCode(__func__, result);
+}
+
+void CurlDownloadRequest::HandleCurlMultiErrorCode(char const* where,
```

Comment:
From reading the name of the function, I'd assume I should only pass failure codes to it (not a return code indicating a success).

How about something like `RaiseErrorIfBadCurlMultiReturnCode`?

Reply:
Why not just `RaiseOnError()`?  I fixed the other function that was also confusingly named. And we might refactor these two classes in the future, but right now I am not sure if they have enough commonality.

Reply:
Sounds good.